rust/xaynet-core/src/mask/scalar.rs

```diff
@@ -0,0 +1,250 @@
+        Ratio::new(numer.into(), denom.into())
+    }
+
+    /// Convenience method for conversion to a non-negative ratio of `BigInt`.
+    pub(crate) fn to_ratio(&self) -> Ratio<BigInt> {
+        let numer = self.0.numer().clone();
+        let denom = self.0.denom().clone();
+        Ratio::new(numer.into(), denom.into())
+    }
+
+    fn from_ratio(r: Ratio<BigInt>) -> anyhow::Result<Self> {
+        let (inumer, idenom) = r.into();
+        let unumer = BigUint::try_from(inumer)?;
+        let udenom = BigUint::try_from(idenom)?;
+        Ok(Self(Ratio::new(unumer, udenom)))
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Could not convert weight {weight} to primitive type {target}")]
+/// Errors related to scalar conversion into primitives.
+pub struct ScalarCastError {
+    weight: Ratio<BigUint>,
+    target: PrimitiveType,
+}
+
+/// An interface for conversion into a primitive value.
+///
+/// This trait is used to convert a [`Scalar`], which has its own internal
+/// representation, into a primitive type ([`f32`], [`f64`], [`i32`], [`i64`]).
+/// The opposite trait is [`FromPrimitive`].
+pub trait IntoPrimitive<P>: Sized {
+    /// Consumes into a converted primitive value.
+    ///
+    /// # Errors
+    /// Returns an error if the conversion fails.
+    fn into_primitive(self) -> Result<P, ScalarCastError>;
+
+    /// Converts to a primitive value.
+    ///
+    /// # Errors
+    /// Returns an error if the conversion fails.
+    fn to_primitive(&self) -> Result<P, ScalarCastError>;
+
+    /// Consumes into a converted primitive value.
+    ///
+    /// # Panics
+    /// Panics if the conversion fails.
+    fn into_primitive_unchecked(self) -> P {
+        self.into_primitive()
+            .expect("conversion to primitive type failed")
+    }
+}
+
+/// An interface for conversion from a primitive value.
+///
+/// This trait is used to obtain a [`Scalar`], which has its own representation,
+/// from a primitive type ([`f32`], [`f64`], [`i32`], [`i64`]). The opposite
+/// trait is [`IntoPrimitive`].
+pub trait FromPrimitive<P: Debug>: Sized {
+    /// Converts from a primitive value.
+    ///
+    /// # Errors
+    /// Returns an error if the conversion fails.
+    fn from_primitive(prim: P) -> Result<Self, PrimitiveCastError<P>>;
+
+    /// Converts from a primitive value.
+    ///
+    /// If a direct conversion cannot be obtained from the primitive value, it is clamped.
+    fn from_primitive_bounded(prim: P) -> Self;
+}
+
+impl IntoPrimitive<i32> for Scalar {
+    fn into_primitive(self) -> Result<i32, ScalarCastError> {
+        let r = self.0;
+        r.to_integer().to_i32().ok_or(ScalarCastError {
+            weight: r,
+            target: PrimitiveType::I32,
+        })
+    }
+
+    fn to_primitive(&self) -> Result<i32, ScalarCastError> {
+        let r = self.0.clone();
+        r.to_integer().to_i32().ok_or(ScalarCastError {
+            weight: r,
+            target: PrimitiveType::I32,
+        })
```

Comment:
code duplication could be reduced:
```suggestion
        self.clone().into_primitive()
```
similarly for the other impls below.